### PR TITLE
CCMSG-961: Elasticsearch Datastreams

### PR DESCRIPTION
**Pull request by amuamushu**
## Problem

Elasticsearch recently release a new feature called Data Streams. 

## Solution
This PR allows the connector to alternatively write to a data stream and is the accumulation of many smaller PRs that have been approved over time.

##### Does this solution apply anywhere else?
- [ ] yes
- [x] no

##### Testing done:
- [x] Unit tests
- [x] Integration tests
- [ ] System tests
- [x] Manual tests

## Release Plan

Releasing on-premise and on-cloud.

Yes, merging into master.
 
Not reverting or rolling back.

### CODE REVIEW

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java [hunk 1]
```diff
@@ -879,6 +986,45 @@ public WriteMethod writeMethod() {
...
+
+      String dataset = (String) value;
+
+      if (dataset.length() > 100) {
+        throw new ConfigException(
+            name, value, "The specified dataset must be no longer than 100 characters."
+        );
+      }
+
+      if (!value.equals(dataset.toLowerCase())) {
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java:1006]:
```suggestion
      if (!dataset.equals(dataset.toLowerCase())) {
```

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java [hunk 2]
```diff
@@ -879,6 +986,45 @@ public WriteMethod writeMethod() {
...
+    public void ensureValid(String name, Object value) {
+      if (value == null) {
+        return;
+      }
+
+      String dataset = (String) value;
+
+      if (dataset.length() > 100) {
+        throw new ConfigException(
+            name, value, "The specified dataset must be no longer than 100 characters."
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java:1002]:
```suggestion
            name, dataset, "The specified dataset must be no longer than 100 characters."
```

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java [hunk 3]
```diff
@@ -879,6 +986,45 @@ public WriteMethod writeMethod() {
...
+        );
+      }
+
+      if (!value.equals(dataset.toLowerCase())) {
+        throw new ConfigException(
+            name, value, "The specified dataset must be in all lowercase."
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java:1008]:
```suggestion
            name, dataset, "The specified dataset must be in all lowercase."
```

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java [hunk 4]
```diff
@@ -879,6 +986,45 @@ public WriteMethod writeMethod() {
...
+        );
+      }
+
+      if (dataset.matches(".*[\\\\\\/\\*\\?\\\"<>| ,#\\-:]+.*")) {
+        throw new ConfigException(
+            name, value,
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java:1014]:
```suggestion
            name, dataset,
```

File: src/main/java/io/confluent/connect/elasticsearch/Validator.java [hunk 5]
```diff
@@ -98,6 +110,7 @@ public Config validate() {
       if (!hasErrors()) {
         // no point if previous configs are invalid
         validateConnection(client);
+        validateVersion(client);
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/Validator.java:113]: i would put this in a separate `if (!hasErrors())` block

File: src/main/java/io/confluent/connect/elasticsearch/Validator.java [hunk 6]
```diff
@@ -264,6 +322,56 @@ private void validateSsl() {
     }
   }
 
+  private void validateVersion(RestHighLevelClient client) {
+    if (!config.isDataStream()) {
+      return;
+    }
+    MainResponse response;
+    try {
+      response = client.info(RequestOptions.DEFAULT);
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/Validator.java:331]: can you please find out what permissions are needed for this API and add them to our documentation

**Review by levzem — APPROVED** [review]
LGTM great work @amuamushu

File: src/main/java/io/confluent/connect/elasticsearch/Validator.java (hunk 6 above)
Comment (amuamushu) [reply src/main/java/io/confluent/connect/elasticsearch/Validator.java:331]: Resolved in 1:1. The connector would need the `monitor` privilege for the ping and info APIs. The document has been updated by the docs team to reflect this ([PR link](https://github.com/confluentinc/docs-kafka-connect-elasticsearch/pull/47))

**Pull request merged**